judgments/forms/fields.py

```diff
@@ -0,0 +1,98 @@
+
+    The expected use case for this is to have two of these fields on a given
+    form, one being date from and one being date to.
+
+    If a day or month is not provided, default to the minimum or maximum
+    date value depending on whether it is to/from.
+
+    If year is missing raise a validation error.
+
+    e.g. - value of only year provided as from 2010, default to 01/01/2010
+         - value of only year provided as to 2010, default to 31/12/2010
+         - value of year and month provded as to 02/2010, default to 28/02/2010
+    """
+
+    DateType = Literal["from", "to"]
+
+    date_type: DateType
+
+    def __init__(self, date_type, **kwargs):
+        self.date_type = date_type
+        # Define one message for all fields.
+        error_messages = {
+            "required": "Enter the day, month and year",
+            "incomplete": "Enter the day, month and year",
+        }
+        # Or define a different message for each field.
+        fields = (
+            forms.CharField(
+                label="Day",
+                error_messages={"incomplete": "Missing Day"},
+                validators=[RegexValidator(r"^[0-9]+$", "Enter a valid date")],
+                required=False,
+            ),
+            forms.CharField(
+                label="Month",
+                error_messages={"incomplete": "Missing Month"},
+                validators=[RegexValidator(r"^[0-9]+$", "Enter a valid month")],
+                required=False,
+            ),
+            forms.CharField(
+                label="Year",
+                error_messages={"incomplete": "Missing year."},
+                validators=[validate_year_is_within_sensible_range],
+                required=True,
+            ),
+        )
+        forms.MultiValueField.__init__(
+            self, error_messages=error_messages, fields=fields, **kwargs
+        )
+
+    def compress(self, data_list):
+        """
+        Convert the values entered into the fields as a ``date``.
+        * Return None if year is not provided, raising an error in validation.
+        * Default to beginning or end of month if day is not provided depending on
+            whether this is a from or to field.
+        * Default to the minimum or maximum month depending on whether this is a
+            from or a to respectively
+        * Default to the minimum month and day if only year is provided and this is a from date,
+            and vice-versa for a to date
+
+        Args:
+            data_list (tuple): a 3-tuple the of values entered into the fields.
+
+        Returns:
+            the ``date`` for the values entered in the day, month and year fields.
+                If any of the field are blank then None is returned.
+
+        """
+        day, month, year = [int(value) if value else None for value in data_list]
+        if not year:
+            return None
+        if self.date_type == "from":
+            if not month:
```

Comment:
Aha, i now see that the "empty day/ month" logic is indeed here, nice work. Maybe we don't need the error message above in that case?

Reply:
Good shout, leftover from cripsy-forms-gds!